src/client.rs

```diff
@@ -0,0 +1,170 @@
+use std::fmt;
+
+use futures::{SinkExt, TryStreamExt};
+use ssh_key::Signature;
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_util::codec::Framed;
+
+use crate::{
+    codec::Codec,
+    proto::{
+        AddIdentity, AddIdentityConstrained, AddSmartcardKeyConstrained, Extension, Identity,
+        ProtoError, RemoveIdentity, Request, Response, SignRequest, SmartcardKey,
+    },
+};
+
+#[derive(Debug)]
+pub struct Client<Stream>
+where
+    Stream: fmt::Debug + AsyncRead + AsyncWrite + Send + Unpin + 'static,
+{
+    adapter: Framed<Stream, Codec<Response, Request>>,
+}
+
+impl<Stream> Client<Stream>
+where
+    Stream: fmt::Debug + AsyncRead + AsyncWrite + Send + Unpin + 'static,
+{
+    pub fn new(socket: Stream) -> Self {
+        let adapter = Framed::new(socket, Codec::default());
+        Self { adapter }
+    }
+}
+
+#[async_trait::async_trait]
+impl<Stream> crate::agent::Session for Client<Stream>
+where
+    Stream: fmt::Debug + AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+{
+    async fn request_identities(&mut self) -> Result<Vec<Identity>, Box<dyn std::error::Error>> {
+        if let Response::IdentitiesAnswer(identities) =
+            self.handle(Request::RequestIdentities).await?
+        {
+            Ok(identities)
+        } else {
+            Err(ProtoError::UnexpectedResponse.into())
```

Comment:
See comment below for handling of specific error responses, but this will also eat up `UnsupportedCommand { command: u8 }` errors, which would be useful to see on the client (i.e., the agent may be sending something nonstandard)

Reply:
:thinking: I don't think it'll eat up `UnsupportedCommand`s since there's `await?` and it should bubble this error up (line 81). The problem with current handling is I think slightly different: it's not clear *what* was the command returned and what was expected.

Or maybe I misunderstood your concern...?

Reply:
Ah yep - missed the `?`! Now I see that this only catches the case where `handle()` returns something that isn't expected. Ignore me!